Add initial implementation of the new updater

computerquip-streamlabs opened the pull request:
This replaces #839. This is the initial patches for the new updater, along with update and release scripts. Release script still needs some testing but given I've spent literally half a day it trying it a handful of times, I think it's better to just throw this out there. Still wouldn't recommend pulling into staging yet until we're sure we're done with electron-updater.

computerquip-streamlabs commented on bin/release.js:
```diff
@@ -60,6 +61,82 @@ function checkEnv(varName) {
+        '--secret-access-key', s3_skey,
+        '--version', new_version,
+        '--release-dir', app_dir,
+        ...args
```
Yeah, I set some defaults to use a particular bucket since I was originally using the scripts directly from the CLI but I'm not sure I like that anymore (easy to mess up). I'm going to remove those defaults I think to remove that risk.

computerquip-streamlabs commented on main.js:
```diff
@@ -313,7 +323,27 @@ if (shouldQuit) {
+      } else {
+        startApp();
+      }
+    });
```
I'm unsure if this does all that I intended but the exported function from the bootstrap module looks like this:
```
module.exports = async (info) => {
    return entry(info).catch((error) => {
        console.log(error.message);
        return Promise.resolve(false);
    });
}
```
The intention here is that if the entry promise is resolved, it's passed back to the application. Entry may also return false as a valid value as well! If an error or exception occurs in entry, it should be caught in the catch handler here and return false instead of passing that exception to the application.

andycreeth replied:
Yup that should do what you want.

computerquip-streamlabs commented on updater/index.html:
```diff
@@ -3,7 +3,8 @@
     <link href="../vendor/roboto.css" rel="stylesheet">
-    <link rel="stylesheet" type="text/css" href="../vendor/fontawesome-free-5.1.0-web/css/all.css" />
+    <link rel="stylesheet" type="text/css" href="../vendor/twitchalerts.css" />
+    <link rel="stylesheet" type="text/css" href="../vendor/font-awesome-4.7.0/css/font-awesome.min.css" />
```
Yeah, I'll remove this. I thought I fixed this when I rebased... I'm not honestly sure where this was pulled in from as I don't recall messing with any CSS myself.

computerquip-streamlabs commented:
I've made a bunch of changes that need to be tested again including but not limited to:
* Switching scripts from 🐍 case to 🐫 case. I need to be on the look out for find-replace errors.
* Switching indentation from 4 space to 2 space indentation.
* Having one-line if-statement blocks prefer being on the same line as the if statement itself or in braces if it's too long.
* A minor bug where I didn't specify `let` or `const` when using a `for...of` loop. 
* Simplifying the release scripts by implicity using the process environment variables for keys and using globally set variables to set the s3-bucket in a more centralized manner.
* I've removed any implicitly default variables in the scripts to remove any risk of accidentally setting the incorrect parameter. I'd rather the script fail than do the wrong thing.
* I've simplified some of the release script logic concerning calling submodules. It's been abstracted away into a function for more consistent behavior and easier to read code. 
* The ability to pass custom arguments to submodules has been removed to simplify those functions since we can just modify the functions anyways. Not sure why I thought that was necessary in retrospect...
* A minor change to how a return value was given in an asynchronous function to be more clear and consistent.
* Changed the s3 link to a cloudfront link. This actually makes me slight nervous since the http library I use is incredibly low-level (like to a ridiculous point) and if redirects are used, I'm not immediately sure if it will automatically handle those. I'll have to test this. 

Most of the code was also put through a formatter which seems to be pretty close to what the general codebase uses. If there are any oddities, let me know.

Still known to do:
* Remove the CSS additions manually since they don't appear to be associated with any particular commit. Must been something to do with the tree I was on at the time I started implementing this.

computerquip-streamlabs commented:
I've squashed all of the commits and fixed the remaining issues. I've run some testing scenarios (the release script is a bit hard to test unfortunately but it works from what I can tell).

Let's have another go at this soon.
